### PR TITLE
Parser_01

### PR DESCRIPTION
**Pull request by kacperaaa**

### CODE REVIEW

File: src/Parser.h [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+#include "http_get.h"
+#include "Price_info.h"
+
+
+
+
+class Parser{
+    public:
+        std::vector<Price_info> data;
+        std::vector<Price_info> load (std::string);
```
Comment (mateuszaaa) [inline src/Parser.h]: load method should not modify input string and as input string can be quite big we dont want to copy it - use const ref

File: src/Price_info.h [hunk 2]
```diff
@@ -0,0 +1,20 @@
+#pragma once
+
+#include <string>
+
+struct Price_info{
```
Comment (mateuszaaa) [inline src/Price_info.h]: all classes should be named using one of conventions:
*PriceInfo* (i would suggest that one)
price_info

File: src/main.cpp [hunk 3]
```diff
@@ -1,15 +1,22 @@
...
-
     std::cout << "connecting to market" << std::endl;
 
-    std::string data = get("api.binance.com", 443, "/api/v3/ticker/bookTicker");
-    
-    std::cout << data << std::endl;
+
+    std::string binance_string = get("api.binance.com", 443, "/api/v3/ticker/bookTicker");
+    Parser binance_data;
+    binance_data.load(binance_string);
```
Comment (mateuszaaa) [inline src/main.cpp]: please print all the fond symbol pair names to demonstrate that the application is working correctly

File: test/Parser_test.cpp [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    valut2.bid_quantity = 41.02000000;
+    valut2.ask_price = 0.00422000;
+    valut2.ask_quantity =0.23000000;
+    Price_info valut3;
+    valut3.symbol = "BNBBTC";
+    valut3.bid_price = 0.00230510;
+    valut3.bid_quantity = 1.69000000;
+    valut3.ask_price = 0.00230540;
+    valut3.ask_quantity =0.62000000;
+    output.push_back(valut1);
```
Comment (mateuszaaa) [inline test/Parser_test.cpp]:
```
    output.push_back(valut1);
    output.push_back(valut2);
    output.push_back(valut3);

    EXPECT_EQ(input, output);
```

be smart:
```
EXPECT_EQ(input, std::vector<Price_info>({valut1, valut2, valut3}));
```

File: test/Parser_test.cpp [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    output.push_back(valut2);
+    output.push_back(valut3);
+
+    EXPECT_EQ(input, output);
+}
+
+TEST(Parser_test, third_test){
+    std::vector<Price_info> input;
+    Parser string;
+    input = string.load(
```
Comment (mateuszaaa) [inline test/Parser_test.cpp:57]:
```
    input = string.load(
        R"(
            [{""}]
            )");
```

put into single line to make it more readable

File: test/Parser_test.cpp [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        R"(
+            [{"symbol":"ETHBTC","bidPrice":"0.03174400","bidQty":"0.25500000","askPrice":"0.03174900","askQty":"24.00000000"},
+            {"symbol":"LTCBTC","bidPrice":"0.00421900","bidQty":"41.02000000","askPrice":"0.00422000","askQty":"0.23000000"},
+            {"symbol":"BNBBTC","bidPrice":"0.00230510","bidQty":"1.69000000","askPrice":"0.00230540","askQty":"0.62000000"}]
+            )");
+    
+    std::vector<Price_info> output;
+    Price_info valut1;
+    valut1.symbol = "ETHBTC";
```
Comment (mateuszaaa) [inline test/Parser_test.cpp:33]: `valut1` -> `price_info1`
same for the rest

File: src/Parser.h [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+#include <iostream>
+#include <sstream>
+#include <vector>
+#include "http_get.h"
+#include "Price_info.h"
+
+
+
+
```
Comment (mateuszaaa) [inline src/Parser.h:7]: too many empty lines

File: src/Parser.h [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+class Parser{
+    public:
+        std::vector<Price_info> data;
```
Comment (mateuszaaa) [inline src/Parser.h]: seems to be unused - remove it

File: src/Parser.cpp [hunk 9]
```diff
@@ -0,0 +1,69 @@
+#include "Parser.h"
+std::vector<Price_info> Parser::load (std::string data){
+
+    for (std::size_t i=0; i<=data.length(); i++){
+        if(data[i] == '"' or data[i] == '[' or data[i] == '{' or data[i] == ',' or data[i] == ':' or data[i] == '}'|| data[i] == ']' ){
+            data[i] = ' ';
+        }
+        if(data[i]=='s' and data[i+1]=='y' and data[i+2]=='m' and data[i+3]=='b' and data[i+4]=='o' and data[i+5]=='l'){
```
Comment (mateuszaaa) [inline src/Parser.cpp]: here and in above/below cases code is not very readable
```
        if(data[i]=='s' and data[i+1]=='y' and data[i+2]=='m' and data[i+3]=='b' and data[i+4]=='o' and data[i+5]=='l'){
            data[i]=' ';
            data[i+1]=' ';
            data[i+2]=' ';
            data[i+3]=' '; 
            data[i+4]=' ';
            data[i+5]=' ';
        }
```

should be converted to 
```
        if(isSymbolString(data,i)){
           std::string symbol_str("symbol");
           fillWithSpaces(data,i, str.length())
        }
```

same for rest of if cases

File: src/Parser.cpp [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+            data[i+5]=' ';
+        }
+    }
+    std::stringstream stream;
+    stream << data;
+    Price_info temp = {};
+    while (true){
+        if(not (stream >> temp.symbol)){
+            break;
+        }        stream >> temp.bid_price;
```
Comment (mateuszaaa) [inline src/Parser.cpp]: fix formatting

File: src/Parser.cpp [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+        stream >> temp.bid_quantity;
+        stream >> temp.ask_price;
+        stream >> temp.ask_quantity;
+
+        Parser::data.push_back(temp);
```
Comment (mateuszaaa) [inline src/Parser.cpp]: there is  no need for data to be a class member; you can define 
`std::vector<Price_info> result` at the beginning of this function.

**Review by mateuszaaa — COMMENTED** [review]
few comments to address

File: src/Parser.h (hunk 1 above)
Comment (kacperaaa) [reply src/Parser.h]: resolved

File: src/Price_info.h (hunk 2 above)
Comment (kacperaaa) [reply src/Price_info.h]: resolved

File: src/main.cpp (hunk 3 above)
Comment (kacperaaa) [reply src/main.cpp]: resolved

File: test/Parser_test.cpp (hunk 4 above)
Comment (kacperaaa) [reply test/Parser_test.cpp]: resolved

File: test/Parser_test.cpp (hunk 5 above)
Comment (kacperaaa) [reply test/Parser_test.cpp:57]: resolved

File: src/Parser.h (hunk 7 above)
Comment (kacperaaa) [reply src/Parser.h:7]: resolved

File: src/Parser.h (hunk 8 above)
Comment (kacperaaa) [reply src/Parser.h]: resolved

File: src/Parser.cpp (hunk 9 above)
Comment (kacperaaa) [reply src/Parser.cpp]: resolved

File: src/Parser.cpp (hunk 10 above)
Comment (kacperaaa) [reply src/Parser.cpp]: resolved

File: src/Parser.cpp (hunk 11 above)
Comment (kacperaaa) [reply src/Parser.cpp]: resolved

**Pull request merged**